### PR TITLE
Support ValueChoice as depth in Repeat

### PR DESCRIPTION
**Pull request by ultmaster**
### Description ###

Repeat can previously accept a fixed depth, or a range of depth.
Added in this PR is a support of ValueChoice as depth.

One prerequisite is to support ``all_options()`` for value choice, which is a iterable of all the possibilities (combinations) of a ValueChoice.

### Checklist ###
  - [x] test case
  - [x] doc

### CODE REVIEW

**Comment by QuanluZhang** [comment]
seems not necessary, though maybe better to have

File: nni/retiarii/converter/graph_gen.py [hunk 1]
```diff
@@ -660,8 +660,9 @@ def _convert_module(self, script_module, module, module_name, module_python_name
             attrs = {
                 'mutation': 'repeat',
                 'label': module.label,
+                'depth': module.depth_choice,
+                'max_depth': module.max_depth,
```
Comment (QuanluZhang) [inline nni/retiarii/converter/graph_gen.py:664]: what is the difference between depth and max_depth?

Comment (ultmaster) [reply nni/retiarii/converter/graph_gen.py:664]: `depth` is the chosen depth, `max_depth` and `min_depth` is a depth range.

File: nni/retiarii/nn/pytorch/component.py [hunk 2]
```diff
@@ -57,15 +69,32 @@ def __init__(self,
...
+            self.min_depth = depth if isinstance(depth, int) else depth[0]
+            self.max_depth = depth if isinstance(depth, int) else depth[1]
+            self.depth_choice = ValueChoice(list(range(self.min_depth, self.max_depth + 1)), label=label)
+        elif isinstance(depth, int):
+            self.min_depth = self.max_depth = depth
+            self.depth_choice = depth
+        else:
+            raise TypeError(f'Unsupported "depth" type: {type(depth)}')
         assert self.max_depth >= self.min_depth > 0
         self.blocks = nn.ModuleList(self._replicate_and_instantiate(blocks, self.max_depth))
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/component.py:93]: by the way, why must be max_depth here?

File: nni/retiarii/nn/pytorch/mutator.py [hunk 3]
```diff
@@ -144,14 +144,13 @@ def _retrieve_chain_from_graph(self, graph: Graph) -> List[Node]:
...
-        max_depth = self.nodes[0].operation.parameters['max_depth']
-        if min_depth < max_depth:
-            chosen_depth = self.choice(list(range(min_depth, max_depth + 1)))
         for node in self.nodes:
             # the logic here is similar to layer choice. We find cell attached to each node.
             target: Graph = model.graphs[node.operation.cell_name]
             chain = self._retrieve_chain_from_graph(target)
+            # and we get the chosen depth (by value choice)
+            node_in_model = model.get_node_by_name(node.name)
+            chosen_depth = node_in_model.operation.parameters['depth']
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/mutator.py:153]: what is the type of `chosen_depth`? int or valuechoice?

Comment (QuanluZhang) [reply nni/retiarii/nn/pytorch/mutator.py:153]: which part decides the chosen depth?

File: test/ut/retiarii/test_highlevel_apis.py [hunk 4]
```diff
@@ -578,14 +578,39 @@ def forward(self, x):
...
+    def test_repeat_static(self):
+        class AddOne(nn.Module):
+            def forward(self, x):
+                return x + 1
+
+        @model_wrapper
+        class Net(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.block = nn.Repeat(lambda index: nn.LayerChoice([AddOne(), nn.Identity()]), 4)
```
Comment (QuanluZhang) [inline test/ut/retiarii/test_highlevel_apis.py:596]: is there valuechoice tested?

**Comment by matluster** [comment]
> seems not necessary, though maybe better to have

There have been already use cases in #4524 and #4551.

What I'm trying to say, is that, I didn't "imagine" that people could be using value choice in repeat. They are actually using it "for real".

File: test/ut/retiarii/test_highlevel_apis.py (hunk 4 above)
Comment (matluster) [reply test/ut/retiarii/test_highlevel_apis.py:598]: oops, missed a test.

Thanks for pointing this out. It also helped me catch a bug. :(

File: nni/retiarii/nn/pytorch/component.py (hunk 2 above)
Comment (matluster) [reply nni/retiarii/nn/pytorch/component.py:93]: `max_depth` is used to initialize "enough" blocks when super-net is created. Useful in weight sharing.

File: nni/retiarii/nn/pytorch/mutator.py (hunk 3 above)
Comment (matluster) [reply nni/retiarii/nn/pytorch/mutator.py:153]: It's originally a ValueChoice, but it has already been mutated by a ParameterChoiceMutator here.

**Review by QuanluZhang — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**